paper.tex

```diff
@@ -522,30 +515,69 @@ \subsection{GPU: Execution Time} \label{subsection:gpu-results}
+
+\begin{figure}
+    \includegraphics[scale=0.66]{tag_monoid_gpu_timings}
+    \caption{Mean GPU execution time of the tag monoid parallel prefix sums with increasing number of input path segments, measured on Apple M1 Max and NVIDIA RTX 4090. The figure also shows the execution time of the stroke expansion and flattening kernel. The y-axis is shown in logarithmic scale to highlight the similar scaling trend on both GPUs despite the order-of-magnitude difference in performance.}
+    \Description{Mean GPU execution time of the tag monoid kernel}
+    \label{fig:tag-monoid-gpu-timings}
+\end{figure}
+
+\begin{table}
+    \caption{Comparison of CPU encoding time with and without size estimation on the largest scenes. All measurements were computed on an Apple M1 Max. The additional cost is negligible when the scene size is modest but tangible with increased complexity.}
+    \begin{tabular}{|l|c|c|}
+    \hline
+        \multicolumn{1}{|c|}{\textbf{Test}} & \textbf{Encoding} & \textbf{Encoding + Estimation} \\
+    \hline
+        roads.svg  & 398.56 $\mu$s & 618.24 $\mu$s \\
+        waves.svg  & 197.53 $\mu$s & 418.96 $\mu$s \\
+        mmark-70k  & 2.69 ms       & 6.89 ms       \\
+        mmark-120k & 4.28 ms       & 11.59 ms      \\
+    \hline
+    \end{tabular}
+    \label{table:bump-estimate-cost}
+\end{table}
+
+\begin{table}
+    \caption{The estimated and actual \emph{line soup} buffer sizes for scenes rendered at 2088x1600 resolution. Our estimation uses Wang's formula combined with a scaling heuristic based on the linewidth. In the worst case, our buffer size estimator overshoots by a factor ~4.}
+    \begin{tabular}{|l|c|c|}
+    \hline
+        \multicolumn{1}{|c|}{\textbf{Test}} & Actual Size & Estimated Size \\
+    \hline
+        roads.svg  & 958.78 KB & 2.45 MB    \\
+        waves.svg  & 9.37 MB   & 10.44 MB   \\
+        mmark-70k  & 23.81 MB  & 100.79 MB  \\
+        mmark-120k & 40.86 MB  & 173.56 MB  \\
+    \hline
+    \end{tabular}
+    \label{table:bump-estimate-mem}
+\end{table}
 
 \section{Future work}
 
 \begin{itemize}
     \item The GPU implementation sequentializes some work that could be parallel, and is thus not as well load-balanced as it might be. An appealing future direction is to split the pipeline into separate stages, executed by the GPU as nodes in a work graph\cite{Patel2024}. This structure is appealing because load balancing is done by hardware.
 
+    \item The pre-allocation requirement for bump-allocated line soup buffer is a limitation due to today's graphics APIs. A more accurate and optimized buffer size estimation heuristic is considered future work. Another interesting future exploration is to employ newer GPU API features, such as mesh shaders and work graphs, for GPU-driven scheduling of the compute workload under a bounded memory footprint.
```

Comment:
My main hesitation is that I don't see either mesh shaders or work graphs actually solving the GPC problem. I'd rewrite this:

\item Today's graphics APIs do not support dynamic reallocation from within a shader, thus force us to pre-allocate the bump-allocated line soup buffer. A more accurate and optimized buffer size estimation heuristic is considered future work. It may also be interesting to explore whether graphics APIs can be extended to facilitate GPU-driven scheduling of the compute workload under a bounded memory footprint.

I think it's also worth stating, probably earlier, that CUDA *can* reallocate, making it clear that this really is a graphics API specific problem.

Reply:
Done.